pingcap/tiflow
# kv (ticdc): fix kvClient reconnection downhill loop

### What problem does this PR solve?


Issue Number: close #10239 

### What is changed and how it works?

1. Add an ID to the `eventFeedStream` struct to identify a stream and prevent it from being deleted unexpectedly.
2. Bind the cancel function of a gRPC stream to its `eventFeedStream` to prevent the stream from being canceled unexpectedly.
3. Reduce the number of calls to `s.deleteStream` to only once to prevent the stream from being deleted unexpectedly.

### Check List 

#### Tests 


 - Manual test (add detailed scripts or steps below)

1. Add hard code `time.sleep(5 * time.second)` before calling `s.deleteStream` in both unfixed and fixed version of code. 
2. Deploy a TiDB cluster with 3 TiKV nodes and 1 TiCDC server, create a changefeed. 
3. Patch the changed cdc binary, and restart a TiKV node randomly and observe the changefeed's lag. 

unfixed cdc:
![img_v3_027f_9b5e20b7-2aab-44b7-8884-fc29d2797beg](https://github.com/pingcap/tiflow/assets/20351731/013bfffe-6593-4260-8167-7b80ef009fe5)


fixed cdc:
![image](https://github.com/pingcap/tiflow/assets/20351731/363f2cae-1aef-43ac-8022-37e92c4e82d1)
 
From the above graphs, it is evident that in the unfixed CDC, the lag of resolvedTs can exceed 12 minutes when a TiKV node is restarted. However, in the fixed CDC, the increase in resolvedTs is limited to a maximum of 35 seconds. This demonstrates the effectiveness of the fix.

Moreover, when the hard-coded `time.sleep(5 * time.second)` is removed and the fixed version of CDC is tested again, the lag becomes even smaller:
![image](https://github.com/pingcap/tiflow/assets/20351731/e753d9a9-77ef-4f4a-9239-60f8e3c6e799)



#### Questions 

##### Will it cause performance regression or break compatibility?

##### Do you need to update user documentation, design documentation or monitoring documentation?

### Release note 

```release-note
Fix a bug in kv client that could cause an increase in changefeed lag when TiKV is restarted.
```
